blog/zh-cn/rpc-introduction.md

```diff
@@ -0,0 +1,132 @@
+---
+title: 浅谈 RPC
+keywords: RPC, Dubbo, Spring Cloud
+description: RPC-远程过程调用，它是一种通过网络从远程计算机程序上请求服务，而不需要了解底层网络技术的协议。
+---
+
+近几年随着微服务化项目的崛起，逐渐成为许多公司中大型分布式系统架构的主流方式，而今天所说的 RPC 在这其中扮演着至关重要的角色。随着这段日子公司项目微服务化的演进，发现在日常开发中都在隐式或显式的使用 RPC，一些刚刚接触 RPC 的小伙伴会感觉无所适从，而一些入行多年的老手虽然使用 RPC 经验丰富，但有些对其原理也一知半解，缺乏对原理的深入理解，往往也会造成开发中的一些误用。
+
+## 什么是RPC？
+
+RPC（Remote Procedure Call）—远程过程调用，它是一种通过网络从远程计算机程序上请求服务，而不需要了解底层网络技术的协议。也就是说两台服务器A，B，一个应用部署在A服务器上，想要调用B服务器上应用提供的方法，由于不在一个内存空间，不能直接调用，需要通过网络来表达调用的语义和传达调用的数据。
+
+RPC协议假定某些传输协议的存在，如TCP或UDP，为通信程序之间携带信息数据。在OSI网络通信模型中，RPC跨越了传输层和应用层。RPC使得开发包括网络分布式多程序在内的应用程序更加容易。现在业界有很多开源的优秀 RPC 框架，例如 Spring Cloud、Dubbo、Thrift 等。
+
+## RPC 起源
+
+RPC 这个概念术语在上世纪 80 年代由 **Bruce Jay Nelson** 提出。这里我们追溯下当初开发 RPC 的原动机是什么？在 Nelson 的论文 "Implementing Remote Procedure Calls" 中他提到了几点：
+* 简单：RPC 概念的语义十分清晰和简单，这样建立分布式计算就更容易。
+* 高效：过程调用看起来十分简单而且高效。
+* 通用：在单机计算中过程往往是不同算法部分间最重要的通信机制。 
+
+通俗一点说，就是一般程序员对于本地的过程调用很熟悉，那么我们把 RPC 作成和本地调用完全类似，那么就更容易被接受，使用起来毫无障碍。Nelson 的论文发表于 30 年前，其观点今天看来确实高瞻远瞩，今天我们使用的 RPC 框架基本就是按这个目标来实现的。
+
+## RPC 结构
+
+Nelson 的论文中指出实现 RPC 的程序包括 5 个部分：
+1. User
+2. User-stub
+3. RPCRuntime
+4. Server-stub
+5. Server
+
+![RPC结构](https://cdn.nlark.com/yuque/0/2019/png/338441/1564718953830-88ebb4b3-9346-4153-a497-de5fba6566b2.png)
+
+这里 user 就是 client 端，当 user 想发起一个远程调用时，它实际是通过本地调用 user-stub。user-stub 负责将调用的接口、方法和参数通过约定的协议规范进行编码并通过本地的 RPCRuntime 实例传输到远端的实例。远端 RPCRuntime 实例收到请求后交给 server-stub 进行解码后发起本地端调用，调用结果再返回给 user 端。
+
+以上是粗粒度的 RPC 实现概念结构，接下来我们进一步细化它应该由哪些组件构成，如下图所示。
+
+![RPC 结构拆解](https://cdn.nlark.com/yuque/0/2019/png/338441/1564725591452-834d8ea4-2491-439c-87e2-ef9dfac1ef3e.png)
+
+RPC 服务方通过 RpcServer 去导出（export）远程接口方法，而客户方通过 RpcClient 去引入（import）远程接口方法。客户方像调用本地方法一样去调用远程接口方法，RPC 框架提供接口的代理实现，实际的调用将委托给代理RpcProxy 。代理封装调用信息并将调用转交给RpcInvoker 去实际执行。在客户端的RpcInvoker 通过连接器RpcConnector 去维持与服务端的通道RpcChannel，并使用RpcProtocol 执行协议编码（encode）并将编码后的请求消息通过通道发送给服务方。
+
+RPC 服务端接收器 RpcAcceptor 接收客户端的调用请求，同样使用RpcProtocol 执行协议解码（decode）。解码后的调用信息传递给RpcProcessor 去控制处理调用过程，最后再委托调用给RpcInvoker 去实际执行并返回调用结果。如下是各个部分的详细职责：
+
+```
+1. RpcServer  
+
+   负责导出（export）远程接口  
+
+2. RpcClient  
+
+   负责导入（import）远程接口的代理实现  
+
+3. RpcProxy  
+
+   远程接口的代理实现  
+
+4. RpcInvoker  
+
+   客户方实现：负责编码调用信息和发送调用请求到服务方并等待调用结果返回  
+
+   服务方实现：负责调用服务端接口的具体实现并返回调用结果  
+
+5. RpcProtocol  
+
+   负责协议编/解码  
+
+6. RpcConnector  
+
+   负责维持客户方和服务方的连接通道和发送数据到服务方  
+
+7. RpcAcceptor  
+
+   负责接收客户方请求并返回请求结果  
+
+8. RpcProcessor  
+
+   负责在服务方控制调用过程，包括管理调用线程池、超时时间等  
+
+9. RpcChannel  
+
+   数据传输通道 
+```
+
+## RPC 工作原理
+
+RPC的设计由Client，Client stub，Network ，Server stub，Server构成。 其中Client就是用来调用服务的，Cient stub是用来把调用的方法和参数序列化的（因为要在网络中传输，必须要把对象转变成字节），Network用来传输这些信息到Server stub， Server stub用来把这些信息反序列化的，Server就是服务的提供者，最终调用的就是Server提供的方法。
+
+![RPC工作原理](https://cdn.nlark.com/yuque/0/2019/jpeg/338441/1564724156391-2e5310f9-d646-47ec-ae3b-d843a4e61016.jpeg)
```

Comment:
It is recommended to place pictures locally, and if you refer to other people's pictures, it is recommended that the source be placed at the end of the article as a comment, too.

Reply:
> It is recommended to place pictures locally, and if you refer to other people's pictures, it is recommended that the source be placed at the end of the article as a comment, too.

I understand what you mean, put the pictures of the articles on ` images/blog` directory, then stated source images.